Min # Store step7

undervineg opened the pull request.

Commit:
[설정] SwiftLint 적용

Commit:
[추가] 메인 json 파일 추가

Commit:
[추가] 테이블 셀 내부 데이터를 저장할 구조체 추가

Decodable을 채택하여 JSON 데이터를 decode 할 수 있도록 함.

Commit:
[추가] 셀 내부 콘텐츠뷰들의 스타일을 지정하기 위한 상수값 일괄 저장

Commit:
[추가] json 파일을 Data로 변환한 후, 사용가능한 특정 모델타입으로 변환하는 클래스 추가

Commit:
[추가] 데이터가 들어갈 커스텀 셀 뷰 클래스 구현

Commit:
[추가] 뱃지 라벨 1개를 표현하기 위한 UILabel 클래스 작성

Commit:
[추가] 여러 개 삽입될 수도 있는 뱃지를 가로로 균등하게 배치하기 위한 스택뷰 클래스 작성

뱃지의 intrinsic 크기에 맞춰 스택뷰의 크기를 조정할 수 있도록 구현

Commit:
정가, 할인가 데이터가 겹치지 않고 나란히 보일 수 있도록 스택뷰 클래스 구현

Commit:
[추가] 뷰 클래스 부가설정을 위한 프로토콜 작성

Commit:
기본 UIColor에 없는 색상을 정의

Commit:
정가 위에 취소선을 긋고, 할인가의 숫자부분을 강조하기 위한 프로퍼티 추가

Commit:
[추가] 데이터를 뷰에 추가하여 테이블 뷰 완성

Commit:
스토리보드에 테이블뷰의 프로토타입 셀 구성

Commit:
스토리보드에 테이블뷰의 프로토타입 셀 구성

Commit:
[추가] 뷰 설명 추가

Commit:
[정리] 1단계 구현 내용 정리

Commit:
[추가] 2단계 구현 내용 추가

Commit:
[수정] 요구사항에 맞게 제약조건 수정

뱃지 컨테이너의 widthAnchor, heightAnchor priority를 낮추어 가변적으로 만듦.
뱃지 컨테이너의 hugging, compression resistance priority를 높여 콘텐츠 사이즈에 딱 맞게 설정되도록 수정.

Commit:
[변경] 뱃지 라벨의 inset 값 수정

Commit:
[추가] 셀 재사용을 위해 리셋 함수 추가

+ 콘텐츠에 딱 맞추도록 수평/수직 Hugging priority와 compression resistance priority를 높임

Commit:
[추가] 셀 재사용을 위해 하위 뷰 제거 함수 추가

+ 기존 resizeContainer() 메소드 제거

Commit:
셀 재사용 시 뱃지 컨테이너 초기화하도록 prepareForReuse() 메소드 오버라이드

Commit:
행 높이 메소드 오버라이드

Commit:
[추가] 코드 내용 추가

Commit:
[추가] 섹션별 제목, 부제목, 셀 데이터를 담을 구조체 추가

Commit:
[추가] 섹션별 헤더 뷰 작성

Commit:
[추가] 섹션별 데이터 삽입 및 헤더 추가

Commit:
파일 추가

Commit:
[추가] Status bar가 투명해서 콘텐츠가 비쳐보이는 것을 방지하기 위해 색상 설정

Commit:
[수정] 위치 조정

Commit:
일부 스타일값 변경

Commit:
[삭제] 필요없는 메소드 제거

Commit:
뱃지 라벨에서 Inset 라벨을 따로 떼어낸 후, 상속. 다른 뷰에도 Inset 라벨 적용 위함

Commit:
[추가] 3단계 구현 내용 추가

Commit:
[추가] 셀 클릭 시, 제목 및 할인 가격 토스트

Commit:
Toaster 패키지 설치

Commit:
[추가] 4단계 구현 내용 추가

Commit:
띄어쓰기 수정

Commit:
[제거] 이전에 추가했던 항목 중 불필요한 항목 제거

Commit:
[추가] 서버에서 데이터를 받아오는  메소드 추가

Commit:
[수정] 파라미터 라벨 수정

Commit:
[추가] 서버로부터 받은 데이터를 처리하는 메소드 추가

서버에서 받은 데이터를 Section 구조체로 변환하여 로컬 데이터 배열에 삽입한 후, 메인 쓰레드에서 테이블 뷰 데이터 리로드.

Commit:
[추가] URLSession.shared 객체를 사용하지 않는 다운로드 메소드 추가

Commit:
파일 추가/삭제

Commit:
[추가] 서버에서 파일을 받아와 디코딩하는 함수 추가

Commit:
[수정] 프로토타입 셀의 Identifier 변경

Commit:
[추가] 헤더 스타일 추가(테두리)

Commit:
[수정] xib 파일 제거. 코드로만 작성

Commit:
[수정] 오토레이아웃 제약조건 설정하는 내부 메소드를 UIView extension 메소드로 대체

Commit:
configure(label:style:) 메소드를 UIView의 extension 메소드로 대체

Commit:
[추가] Reusable 뷰의 reuseIdentifier를 간편하게 사용하기 위한 extension 추가

Commit:
[수정] Section 클래스의 기존 title, subtitle 제거. type 프로퍼티로 유추가능

Commit:
[추가] 로컬서버/리모트서버를 구분하기 위한 열거형 추가

Commit:
[추가] 섹션당 데이터를 로드하여 비동기로 업데이트하는 로직 추가

Commit:
[추가] 5단계 구현 내용 추가

Commit:
[수정] 다운로드만 관장하도록 변경. (디코드 로직 삭제)

Commit:
[추가] 캐시 디렉토리에 데이터를 저장/불러오기 할 수 있는 클래스 작성

Commit:
[이동] TableSection 열거형을 새 파일로 분리

Commit:
[추가] 썸네일을 비동기로 다운로드하여 저장하는 클래스 추가

StoreItem들을 다운로드 받아 디코딩 시, 썸네일 url을 통해 각 StoreItem 마다 새 쓰레드를 생성하여 다운로드.
StoreItem 구조체에 썸네일을 저장하는 변수 추가. (Thumbnail 타입)

Commit:
뷰컨트롤러에서 셀에 데이터 삽입하는 부분을 ItemCell에 넣었다가 다시 뷰컨트롤러로 옮김

Commit:
[추가] 썸네일 이미지 비동기 표시

섹션당 [StoreItem]이 다운로드될 때 썸네일이 비동기로 다운로드 되므로 해당 로직은 Thumbnail 클래스에 있음.
썸네일 이미지를 비동기로 표시하기 위해 tableView(cellForRowAt:) 메소드에 DispatcheQueue.main.async 블록 추가.
+ loadItemsFromAPI()에서 섹션별 데이터가 다운로드 된 후 섹션별 삽입 추가.

Commit:
KVO 방식으로 구현 테스트

Commit:
[추가] 비동기로 이미지 표시

+ 섹션별 데이터 로드 후 insert 하도록 변경 (tableView.reloadData 제거)

Commit:
6단계 구현 내용 추가

Commit:
파일 추가 및 제거

Commit:
[추가] Detail 화면 및 제약조건 추가

Commit:
[변경] 일부 프로퍼티, 파라미터, 함수 이름 변경

Commit:
[분리] 파일 분리

Commit:
[수정] Thumbnail 생성자 앞에 try 추가

Commit:
[추가] 데이터를 뷰에 추가하는 로직 작성

Commit:
[추가] POST 요청 가능한 메소드 추가

Commit:
[추가] Detail 뷰컨트롤러로 데이터 전달 및 화면전환 로직 추가

Commit:
[추가] 데이터 로드 및 주문하기 버튼 클릭 이벤트 추가

godrm commented on StoreApp/StoreApp/DetailViewController.swift:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    private func loadItemsFromAPI() {
+        guard let url = URL(string: "http://crong.codesquad.kr:8080/woowa/detail/"+detailHash!) else { return }
```
하드코딩한 URL 주소를 관리하는 방법은 뭐가 좋을까요?
항상 숫자나 문자열도 의미있는 코드로 표현해보세요.

godrm commented on StoreApp/StoreApp/DetailViewController.swift:
```diff
@@ -0,0 +1,59 @@
+            let contents = try Data(contentsOf: url)
+            self.items = try JSONDecoder().decode(ItemDetail.self, from: contents)
+        } catch {
+            NSLog(error.localizedDescription)
```
여기서 catch가 되고 나면 추가적인 흐름 제어는 필요가 없나요?
아래도 내려가서 presentView()를 실행해도 되는건가요?

godrm commented on StoreApp/StoreApp/DetailViewController.swift:
```diff
@@ -0,0 +1,59 @@
+        let price = String(describing: items!.data.prices.first!)
+        let menu = String(describing: items!.data.productDescription)
+        let payload = ["text": "- 주문자: Min,\n - 가격: \(price),\n - 메뉴: \(menu)"]
+        do {
```
여기 있는 문자열도 위와 같은 방법으로 고민해 보세요

godrm commented on StoreApp/StoreApp/DetailViewController.swift:
```diff
@@ -0,0 +1,59 @@
+            let payloadData = try JSONSerialization.data(withJSONObject: payload, options: .init(rawValue: 0))
+            Downloader.post(to: urlString, with: payloadData)
+        } catch {
+            NSLog(error.localizedDescription)
```
예외적인 흐름을 어떻게 처리하면 좋을까 고민해보세요.

godrm commented on StoreApp/StoreApp/Helper/CacheStorage.swift:
```diff
@@ -0,0 +1,32 @@
+
+    static func save(_ fileName: String, _ data: Data?) {
+        let fileUrl = cacheUrl.appendingPathComponent(fileName)
+        FileManager.default.createFile(atPath: fileUrl.path, contents: data, attributes: nil)
```
기존에 파일이 있으면 어떻게 하나요?

godrm replied:
FileManager 와 달리 Data.write 메소드를 활용하는 것과 어떤 차이가 있을까요?
https://developer.apple.com/documentation/foundation/data/1779858-write

undervineg replied:
음.. doc을 읽어보니까 FileManager는 파일 단위를 관리하기 때문에 파일에 데이터를 붙여쓰거나 하는 건 없어보이네요. FileManager의 createFile은 파일을 아예 새로 생성하는 반면, Data.write은 파일 내 데이터에 새 데이터를 붙일 수 있는 차이가 있는 것 같아요..!

godrm commented on StoreApp/StoreApp/Helper/Downloader.swift:
```diff
@@ -0,0 +1,57 @@
+        request.httpMethod = "POST"
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        request.setValue("application/json", forHTTPHeaderField: "Accept")
+        request.httpBody = payload
```
여기서 사용하는 문자열 상수도 의미있는 코드 단위로 바꾸세요.

godrm commented on StoreApp/StoreApp/Helper/Downloader.swift:
```diff
@@ -0,0 +1,57 @@
+        DispatchQueue.global().async(execute: {
+            if let nsData = NSData(contentsOf: url) {
+                let data = Data(referencing: nsData)
+                completionHandler(.success(data))
```
Data에도 contentsOf: 가 있을텐데요?

godrm commented on StoreApp/StoreApp/Model/Section.swift:
```diff
@@ -0,0 +1,25 @@
+
+    init(type: TableSection, cell: [StoreItem]) {
+        self.type = type
+        self.cells = cell
```
cells와 cell 헷갈리지 않나요? (나만 그런가?)

godrm commented on StoreApp/StoreApp/Model/Server.swift:
```diff
@@ -0,0 +1,55 @@
+
+    var side: String {
+        return "http://crong.codesquad.kr:8080/woowa/side"
+    }
```
여기 네트워크 주소를 관리하는 것은 괜찮네요. 
근데 var가 아니라 그냥 let 이어도 되지 않을까요?

godrm commented on StoreApp/StoreApp/Model/Thumbnail.swift:
```diff
@@ -0,0 +1,47 @@
+                    self.image = UIImage(data: data)
+                case .failure(let error):
+                    if case let DownloadError.downloadFail(message: errorMessage) = error {
+                        NSLog(errorMessage)
```
스위프트에서는 NSLog() 대신에 print()를 쓰는 편입니다. 이건 오브젝티브-C 스타일이네요.
그리고 실패할 경우... 로그만 찍으면 끝인가요? 개선할 방법이 있을까요?

godrm commented on StoreApp/StoreApp/Supporting Files/AppDelegate.swift:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    func applicationWillTerminate(_ application: UIApplication) {
+    }
```
여기 구현안 할 메소드들은 그냥 지워도 되요.

godrm commented on StoreApp/StoreApp/View/BadgesContainer.swift:
```diff
@@ -0,0 +1,39 @@
+
+    func removeAllBadges() {
+        self.arrangedSubviews.forEach {
+            guard let label = $0 as? BadgeLabel else { return }
```
guard let 대신에 where 구문도 가능하지 않을까요?

godrm commented on StoreApp/StoreApp/View/DetailView.swift:
```diff
@@ -0,0 +1,88 @@
+
+protocol DetailViewDelegate: class {
+    func orderButtonDidTapped()
+}
```
이런 프로토콜 선언은 찾기 쉽도록 파일 윗 부분에 표시하는 관례가 있습니다.

godrm commented on StoreApp/StoreApp/View/DetailView.swift:
```diff
@@ -0,0 +1,88 @@
+
+    private func setThumbnailScrollView() {
+        guard let items = items else { return }
+        layoutIfNeeded()
```
layoutIfNeeded(), draw(), viewDidLoad() 이런 메소드들은 서브클래스를 만들도 구현하는 메소드입니다.
그래서 직접 호출하기 보다는 setNeedLayout()을 통해서 다시 갱신하도록 유도합니다. 간혹 곧바로 모든 레이아웃을 갱신하기 위한 용도로 사용하기도 합니다. 
이 글을 읽어보시고 본인 의도가 무엇인지 확인해보세요. `https://medium.com/@abhimuralidharan/ios-swift-setneedslayout-vs-layoutifneeded-vs-layoutsubviews-5a2b486da31c`

godrm commented on StoreApp/StoreApp/View/DetailView.swift:
```diff
@@ -0,0 +1,88 @@
+        setThumbnailScrollView()
+        setInfoLabelsContainer()
+        setDetailSection()
+        setContentSize()
```
뷰를 처음에 설정하는 메소드들은 set-보다는 configure- 같은 표현을 많이 씁니다.

godrm commented on StoreApp/StoreApp/ViewController.swift:
```diff
@@ -0,0 +1,107 @@
+        tableView.estimatedRowHeight = 40
+        tableView.estimatedSectionHeaderHeight = 40
+        tableView.rowHeight = UITableViewAutomaticDimension
+        tableView.sectionHeaderHeight = UITableViewAutomaticDimension
```
이런 override 함수는 반드시 super.xxxx 형태를 빼먹지 않도록 하세요.

godrm commented on StoreApp/StoreApp/ViewController.swift:
```diff
@@ -0,0 +1,107 @@
+                    }
+                })
+            case .failure(let error): NSLog(error.localizedDescription)
+            }
```
클로저 내부에 또 클로저가 반복되는데 이런 구조를 개선하는 방법은 뭐가 좋을까요

undervineg replied:
함수로 빼내긴 했는데 더 좋은 방법이 딱히 떠오르지 않네요..🤨

godrm commented on StoreApp/StoreApp/ViewController.swift:
```diff
@@ -0,0 +1,107 @@
+//        ToastView.appearance().font = UIFont.boldSystemFont(ofSize: 15)
+//        let toaster = Toast(text: row.title+"\n"+row.salePrice)
+//        toaster.show()
+//    }
```
이 부분을 주석처리 하지말고... 상세화면에서 주문 완료한 다음에 여기로 돌아오면 toast를 띄우도록 개선해보세요.

godrm commented on StoreApp/StoreApp/DetailViewController.swift:
```diff
@@ -0,0 +1,59 @@
+
+    override func viewWillAppear(_ animated: Bool) {
+        self.navigationController?.navigationBar.isHidden = false
+    }
```
override 한 경우 super.xxxx를 꼭 처리하세요.

godrm reviewed (commented):
파일이 많아서 변수이름이나 함수이름처럼 API 설계 가이드라인에 나올 법한 내용은 제외했습니다.
그래도 파일수에 비하면 항목이 적네요. 차츰 개선해보세요.

Commit:
[변경] 서버 주소 프로퍼티를 연산 프로퍼티 var에서 저장 프로퍼티 let으로 변경

Commit:
[수정] NSData를 Data로 변경

Commit:
[변경] 파라미터 라벨을 cell->cells로 변경

Commit:
[제거] 구현 안 한 메소드 제거

Commit:
[변경] DetailViewDelegate 선언 위치 변경

Commit:
[수정] layoutIfNeeded()를 setNeedsLayout()으로 변경

Commit:
[수정] 뷰 초기 설정 메소드들의 접두사를 set-에서 configure-로 변경

Commit:
[수정] override 메소드에 super. 추가

Commit:
[이동] Constants 파일만 폴더로 묶음

Commit:
[이동] Constants 파일만 폴더로 묶음

Commit:
[추가] HTTP 헤더 및 바디 설정 관련 상수 추가

Commit:
[수정] 하드코딩했던 텍스트를 상수 프로퍼티로 변경

Commit:
[수정] for문에 다운캐스팅 함께 추가

Commit:
[추가] 네트워크 에러 열거형 추가

Commit:
[추가] 에러 발생 시 썸네일을 회색으로 채움

Commit:
[추가] 이미지 로드 등에서 에러 발생 시 알럿 띄움 추가

Commit:
[추가] 에러 발생 시 썸네일을 회색으로 채움

Commit:
[추가] 이전화면으로 전환 시 메소드 실행을 위한 델리게이트 프로토콜 추가

+ 주문 정보를 저장할 구조체 추가

Commit:
[수정] 기존 menu, price 파라미터를 OrderInfo 구조체 타입으로 변경

Commit:
[추가] 상세화면에서 주문 후 메인으로 돌아와 토스트 띄움

Commit:
[추가] 캐시 디렉토리에 파일이 이미 있는 경우, 제거 후 새 파일 추가

Commit:
[추가] 아이템 로드 실패 시 에러 알럿 추가

Commit:
[분리] 이중 클로저의 내부 클로저를 함수로 분리

undervineg commented:
- [x] 하드코딩한 URL 주소, 문자열(payload 등) 관리
- [x] do-catch문의 추가적인 흐름 제어
    - [x] DetailViewController: 알럿 띄움
    - [x] Thumbnail: 회색 이미지 넣음
    - [x] NSLog는 print로 변경
- [x] CacheStorage에서 기존에 파일이 있는 경우, 파일 지우고 새로 생성
- [x] NSData를 Data로 변경
- [x] 파라미터 라벨 Cell -> Cells로 변경
- [x] 서버 주소 프로퍼티를 let으로 변경
- [x] 구현 안 한 메소드 제거
- [x] removeAllBadges()의 guard문을 for case let문으로 합침
- [x] DetailViewDelegate 선언 위치 변경
- [x] layoutIfNeeded()를 setNeedLayout()으로 변경
- [x] (DetailView) set-을 configure-로 변경
- [x]  override 메소드에 super. 메소드 추가
- [x] 클로저 내 클로저 구조 개선: 함수로 빼내긴 했는데.. 마음에 안 듦 ㅠ
- [x] 주문완료 시 viewController로 돌아오면 toast 띄움 추가

Commit:
[수정] 에러메시지 오타 수정

Commit:
[수정] 글로벌 큐 메소드에 qos 설정 파라미터 추가

+ 썸네일 다운로드 시, .userInteractive 로 설정

undervineg commented:
(좀 더 수정할 부분이 있어서 수정중입니다.)

Commit:
[추가] 이미지 로드 직후 델리게이트 메소드를 호출하도록 프로토콜 및 프로퍼티 옵저버 추가

+ DetailViewController에서 썸네일 스크롤 뷰에 들어갈 이미지와 디테일 섹션에 들어갈 이미지를 분간하기 위해 Thumbnail과 비슷한 역할을 하는 DetailImage 클래스 추가

Commit:
[수정] 기존의 loadImage() 메소드 제거 및 Thumbnail, DetailImage 생성자로 대체

Commit:
[추가] 썸네일 및 디테일 이미지 로드 직후 뷰에 표시하는 델리게이트 메소드 구현

Commit:
[추가] 썸네일 및 디테일 이미지를 한 개씩 받아 뷰에 표시하는 메소드 추가

+ 기존에 이미지 데이터를 한꺼번에 받아 뷰에 표시하는 메소드 제거

Commit:
[정리] OrderResultDelegate 프로토콜을 extension으로 뺌

Commit:
[작성중] 이미지 추가

undervineg commented:
DetailViewController에서 이미지 한 장 불러올 때마다 뷰 업데이트 하도록 수정
- AS-IS: 이미지 데이터 한 번에 불러와서 저장 후 for문으로 뷰에 붙이다 보니 화면 로드 느림
- TO-BE: 이미지 로드 직후 뷰에 표시

godrm commented on StoreApp/StoreApp/DetailViewController.swift:
```diff
@@ -47,14 +48,27 @@ class DetailViewController: UIViewController, Reusable, DetailViewDelegate {
+            return
+        })
+        self.present(alert, animated: true, completion: nil)
+    }
```
참고로 Alert는 사용자에게 버튼을 누르는 액션을 강요하기 때문에 좋지 않는 방법입니다. 
더 좋은 방법이 있는지 고민해보고 최후의 수단으로 사용하세요.

undervineg replied:
요즘 앱에서는 에러가 나면 에러 페이지를 보여주는 것 같은데요~ 많이 생각해봤는데 더 좋은 방법이 있을까요?

godrm replied:
문자 왔을때 뜨는 노티 화면을 생각해보세요. Alert으로 하면 그 창을 닫기전에 다른걸 할수가 없지만, 요즘은 상단에서 보여지고 포커스를 잃지않아서 하던 작업을 할 수 있죠.

Commit:
[추가] 에러 발생 시 상단 알림을 위한 라이브러리 설치

Commit:
[변경] 상단 알림 라이브러리 변경

Commit:
[추가] 예외 발생 시 알럿 대신 상단 메시지 추가

Commit:
[분리] 구조체 파일 새로 생성하여 분리

Commit:
[수정] 프로토콜로 Thumbnail과 DetailImage를 같은 타입으로 묶고, 공통 기능을 해당 프로토콜 extension으로 묶음

+ 파일분리 및 코드이동

godrm commented on StoreApp/StoreApp/Model/AsyncPresentable.swift:
```diff
@@ -0,0 +1,45 @@
+protocol AsyncPresentable: class {
+    func loadImage(from urlString: String)
+    var image: UIImage? { get set }
+}
```
이런 구현 방법도 괜찮네요.

godrm commented on StoreApp/StoreApp/Controller/Extensions/UIViewController+Extension.swift:
```diff
@@ -0,0 +1,26 @@
+                ])
+            GSMessage.font = UIFont.boldSystemFont(ofSize: 15)
+            GSMessage.warningBackgroundColor = UIColor.emerald
+        })
```
이렇게 구현하면 메인에서 에러나고 디테일 띄우면 에러 화면이 두 번 뜨지는 않나요? :)

undervineg replied:
두 번 뜨지는 않는데, 제가 적용한 색상으로 나타나지는 않아서, 폰트, 배경색을 위쪽으로 옮겼습니다.

godrm replied:
UIViewController+Extension.swift 에서 ViewController 자체에 확장해놓으면
네트워크가 안되는 상황에서 띄우는 화면마다 에러가 나지 않을까 해서 물어본거였습니다.

undervineg replied:
확인하기 쉽지는 않은데, 아직까지 그런 현상은 없어요. 근데 함수 호출한 부분에서만 발생하는 게 아닌가요?

godrm replied:
화면이 뜰때마다 요청하고 있지 않나요? 그럴꺼라고 생각했어요. 
그런 현상이 없으면 넘어갑시다. 참.. 폰에서 실행해봤나요?

undervineg replied:
네 폰에서 실행해봤어요. 화면 뜰 때마다 요청하지 않구, 일부 에러 날 때 호출하게 만들었어요.

godrm commented:
수고하셨습니다. 
가볍게 댓글로 확인하고 머지합니다.

Commit:
백그라운드 다운로드 시도하다가 지움^^;

Commit:
[수정] 에러 토스트 색상 및 폰트 적용을 show() 상단으로 옮김

Commit:
[추가] 이미지 파일 추가

Commit:
[수정] 네트워크 에러 시 주문하기 버튼 클릭했을 때 item nil 에러 수정
